### PR TITLE
Error Refactor Done

### PR DESCRIPTION
**Pull request by ControlCplusControlV**
Finally Finished the Error Refactor. This catches up on some other TODOs

My git history on this fork may be a bit fucked though 😆 , working on fixing that from my fork. Squashed everything but on my last PR I accidentally squashed all of my commits to rust-verkle into that one.

### CODE REVIEW

File: verkle-trie/src/errors.rs [hunk 1]
```diff
@@ -1,23 +1,17 @@
...
-    PrecomputedPointsFileExists,
-    #[error("Unable to Create Precomputed Points File")]
-    CannotCreatePrecomputedPoints(std::io::Error),
-    #[error("Precomputed Lagrage Points File Couldn't not be found")]
-    PrecomputedPointsNotFound,
-    #[error("Issue opening PrecomputedPointsFile")]
-    FileError(std::io::Error),
 
+#[derive(Debug, Error)]
+pub enum HintError {
```
Comment (kevaundray) [inline verkle-trie/src/errors.rs:8]: Can we just use std::io::Error as its the only variant at the moment and I'm not sure `HintError` is a unambiguous name

File: verkle-trie/src/errors.rs [hunk 2]
```diff
@@ -28,12 +22,26 @@ pub enum VerkleError {
...
     EmptyPrefix,
+}
+
+#[derive(Debug, Error)]
+pub enum ConfigError {
+    #[error("Precomputed Points Exist Already")]
+    PrecomputedPointsFileExists,
+    #[error("Issue opening PrecomputedPointsFile")]
+    FileError(std::io::Error),
+    #[error("Precomputed Lagrage Points File Couldn't not be found")]
```
Comment (kevaundray) [inline verkle-trie/src/errors.rs:35]: typo : Lagrange

File: verkle-trie/src/errors.rs [hunk 3]
```diff
@@ -28,12 +22,26 @@ pub enum VerkleError {
...
+    PrecomputedPointsNotFound,
+    #[error("Serialization Either Failer or Data is Invalid")]
```
Comment (kevaundray) [inline verkle-trie/src/errors.rs:37]: Can you rewrite this, I'm having a bit of trouble parsing it initially.

typo : Failer -> Failure ?

File: verkle-trie/src/proof.rs [hunk 4]
```diff
@@ -67,11 +64,13 @@ impl std::fmt::Display for VerificationHint {
     }
 }
 
+use std::result::Result;
```
Comment (kevaundray) [inline verkle-trie/src/proof.rs:67]: Is this import needed?

Comment (kevaundray) [inline verkle-trie/src/proof.rs:67]: Is this import needed?

Comment (kevaundray) [reply verkle-trie/src/proof.rs:67]: Note above that we were previously importing std::io::Result, whereas std::result::Result should be in the prelude and hence imported automatically

File: verkle-trie/src/proof.rs [hunk 5]
```diff
@@ -117,7 +116,7 @@ impl VerificationHint {
             diff_stem_no_proof,
         })
     }
-    pub fn write<W: Write>(&self, writer: &mut W) -> IOResult<()> {
+    pub fn write<W: Write>(&self, writer: &mut W) -> Result<(), std::io::Error> {
```
Comment (kevaundray) [inline verkle-trie/src/proof.rs:119]: Was this supposed to return a HintError instead of std::io::Error ?

File: verkle-trie/src/proof/prover.rs [hunk 6]
```diff
@@ -14,17 +15,24 @@ use std::collections::BTreeSet;
...
+    }
     let (queries, verification_hint) = create_prover_queries(storage, keys);
 
     // Commitments without duplicates and without the root, (implicitly) sorted by path, since the queries were
     // processed by path order
-    let root_comm = queries
-        .first()
-        .expect("expected to have at least one query. The first query will be against the root")
-        .commitment;
+    let root_query = queries.first();
```
Comment (kevaundray) [inline verkle-trie/src/proof/prover.rs:26]: You can instead do something like:

```rust
let root_query = match queries.first() {
   Some(query) => query,
   None => return Err(ProofCreationError::ExpectedOneQueryAgainstRoot), 
};
```

This means that you don't  need to unwrap explicitly

File: verkle-trie/src/errors.rs (hunk 1 above)
Comment (kevaundray) [reply verkle-trie/src/errors.rs:8]: Hmm okay I see what you are doing here, the enum is better and HintError can be bikeshedded at a later date

**Review by kevaundray — CHANGES_REQUESTED** [review]
Looks good so far :) Left a few comments for review!

File: verkle-trie/src/proof.rs (hunk 5 above)
Comment (ControlCplusControlV) [reply verkle-trie/src/proof.rs:119]: Yes, although I think next PR I'm going to change this a direct implementation of the `std` "write" trait. Tracking in #69

**Review by kevaundray — APPROVED** [review]

**Pull request merged**